Document Apollo3 GraphQL client errors (#7135)
getsentry/sentry-docs: src/platforms/android/configuration/integrations/apollo3.mdx

import com.apollographql.apollo3.ApolloClient;
import io.sentry.apollo3.SentryApolloBuilderExtensionsKt;

> ApolloClient apollo = SentryApolloBuilderExtensionsKt

Author: I just did indentation here, the code already exists, not sure I follow the question.
There's a difference between Kotlin vs Java as well.

> ApolloClient apollo = SentryApolloBuilderExtensionsKt

Reviewer: Oh right, you only need `SentryApolloBuilderExtensionsKt` on Java, and the snippets below are only in Kotlin, sorry for the confusion. Should we add Java snippets for client errors as well?

> ApolloClient apollo = SentryApolloBuilderExtensionsKt

Author: To be honest, I think we should go Kotlin first and only, less to maintain, nowadays IDEs auto convert the code on copy and paste.
Also, the Apollo3 is designed for KMP, Kotlin first, 90%+ won't be using Kotlin anymore most likely.

> ApolloClient apollo = SentryApolloBuilderExtensionsKt

Reviewer: That's fair, we could add a section "you can also use this in Java like this: <snippet>" and the other snippets only in Kotlin. But not necesserily in ths PR

    }
    .build()
```

## GraphQL Client Errors

This feature, once enabled, automatically captures GraphQL client errors, like bad operations and response codes, as error events and reports them to Sentry. The error event will contain the `request` and `response` data, such as `url`, `status_code`, `data` (stringified `query`) and so on.

Sentry will group GraphQL client errors by the `operationName`, `operationType` and `statusCode`, so that you can easily see how many errors happened for a specific `operationName`, `operationType` and `statusCode`.

This feature is opt-in and can be enabled by setting the `captureFailedRequests` option to `true`:

```kotlin
import com.apollographql.apollo3.ApolloClient
import io.sentry.apollo3.sentryTracing

val apollo = ApolloClient.builder()
    .serverUrl("https://your-api-host/")
    .sentryTracing(captureFailedRequests = true)
    .build()
```

> By default, only GraphQL client errors with a response that includes the [errors](https://spec.graphql.org/October2021/#sec-Errors) array are captured as error events.
> 
> GraphQL client errors from every target (`.*` regular expression) are automatically captured, but you can change this behavior by setting the `failedRequestTargets` option with either a regular expression or a plain `String`. A plain string must contain at least one of the items from the list. Plain strings don't have to be full matches, meaning the URL of a request is matched when it contains a string provided through the option.
> 

Reviewer 2: ```suggestion
By default, only GraphQL client errors with a response that includes the [errors](https://spec.graphql.org/October2021/#sec-Errors) array will be captured as error events.

GraphQL client errors from every target (`.*` regular expression) will be automatically captured, but you can change this behavior by setting the `failedRequestTargets` option with either a regular expression or a plain `String`. Plain strings don't have to be full matches – the URL of a request will be matched if it contains a string provided through the option, but a plain string must contain at least one of the items from the list. 

```
From the list of errors? Can we specify the list of what please?

Author: I'm applying the fix for the 1st part, let me know if the 2nd part needs adaption after clarification here https://github.com/getsentry/sentry-docs/pull/7135/files#r1229248641

```kotlin
import com.apollographql.apollo3.ApolloClient
import io.sentry.apollo3.sentryTracing

val apollo = ApolloClient.builder()
    .serverUrl("https://your-api-host/")
    .sentryTracing(captureFailedRequests = true, failedRequestTargets = listOf("myapi.com"))
    .build()
```

> By default, error events won't contain `Headers` and `Cookies`, but you can change this behavior by setting the `sendDefaultPii` option to `true`:

Reviewer 2: ```suggestion
By default, error events won't contain `Headers` or `Cookies`, but you can change this behavior by setting the `sendDefaultPii` option to `true`:
```

```xml {filename:AndroidManifest.xml}
<application>
  <meta-data android:name="io.sentry.send-default-pii" android:value="true" />
</application>
```

> Error events will contain the raw bodies of GraphQL requests and responses, the raw bodies can contain sensitive data, to avoid that, you can do the parametrization of your queries using the [variables](https://spec.graphql.org/October2021/#sec-Language.Variables) field, using parametrized queries, [Relay](https://docs.sentry.io/product/relay) will run [PII Data Scrubbing](https://docs.sentry.io/product/relay/#pii-data-scrubbing) automatically transforming the values into `[Filtered]`.

Reviewer: ```suggestion
Error events will contain the raw bodies of GraphQL requests and responses, which can include sensitive data. To avoid this, parameterize your queries using the [variables](https://spec.graphql.org/October2021/#sec-Language.Variables) field. [Relay](https://docs.sentry.io/product/relay) will then run [PII Data Scrubbing](https://docs.sentry.io/product/relay/#pii-data-scrubbing), automatically transforming the values into `[Filtered]`.
```

Author: Yep, copypasta error, thank you.

Alternatively, you can customize the event and scrub the data yourself.

### Customize or Drop the Error Event

> To customize or drop the error event, you need to do a [manual initialization](/platforms/android/configuration/manual-init/) of the Sentry Android SDK.
> 
> The captured error event can be customized or dropped with a `BeforeSendCallback`:

Reviewer 2: ```suggestion
To customize or drop the error event, you'll need to do a [manual initialization](/platforms/android/configuration/manual-init/) of the Sentry Android SDK. The captured error event can then be customized or dropped with a `BeforeSendCallback`:
```

```kotlin
import io.sentry.android.core.SentryAndroid
import io.sentry.SentryOptions.BeforeSendCallback
import com.apollographql.apollo3.api.http.HttpRequest
import com.apollographql.apollo3.api.http.HttpResponse
import io.sentry.TypeCheckHint.APOLLO_REQUEST
import io.sentry.TypeCheckHint.APOLLO_RESPONSE

SentryAndroid.init(this) { options ->
  // Add a callback that will be used before the event is sent to Sentry.
  // With this callback, you can modify the event or, when returning null, also discard the event.
  options.beforeSend = BeforeSendCallback { event, hint ->
    val request = hint.getAs(APOLLO_REQUEST, HttpRequest::class.java)
    val response = hint.getAs(APOLLO_RESPONSE, HttpResponse::class.java)

    // customize or drop the event
    event
  }
}
```

> ### Automatically Captured GraphQL Client Errors and Manually Captured Errors

Reviewer 2: ```suggestion
### Automatically and Manually Captured GraphQL Client Errors
```

> When `captureFailedRequests` is enabled, some GraphQL client libraries throw unchecked exceptions like `ApolloException` and its implementations. In this case, the error event is captured twice; once by the GraphQL client library and once by the Sentry Android SDK:

Reviewer: ```suggestion
When `captureFailedRequests` is enabled, some GraphQL client libraries throw unchecked exceptions, like `ApolloException` and its implementations. In this case, the error event is captured twice; once by the GraphQL client library and once by the Sentry Android SDK. We recommend you to identify these errors and **not** capture them manually with the `Sentry.captureException` method:
```

```kotlin
import io.sentry.Sentry
import com.apollographql.apollo3.exception.ApolloException

try {
  // If this API call returns the `errors` array, it will be captured as an error event by the `SentryApollo3HttpInterceptor`.
  return apolloClient.query(LaunchDetailsQuery(launchId)).execute()
} catch (e: ApolloException) {
  // Do not manually capture this exception to avoid duplicated error events.
  // Sentry.captureException(e)
}
```

> We recommend you identify those errors and **not** capture them manually with the `Sentry.captureException` method.

Reviewer: `l`: as commented above, I'd move this above the snippet

Reviewer 2: +1
